src/transformers/models/gpt_neo/modeling_gpt_neo.py

```diff
@@ -0,0 +1,946 @@
+        mask = torch.logical_and(visible, torch.less_equal(relative_position, 0)).transpose(-1, -2).unsqueeze(2)
+        return mask
+
+    def _attn(self, q, k, v, causal_mask, head_mask=None, output_attentions=False):
+        # attn
+        attn_weights = torch.matmul(q, k)
+        attn_weights = torch.where(causal_mask, attn_weights, self.masked_bias.to(attn_weights.dtype))
+
+        attn_weights = nn.Softmax(dim=-1)(attn_weights)
+        attn_weights = self.attn_dropout(attn_weights)
+
+        # Mask heads if we want to
+        if head_mask is not None:
+            attn_weights = attn_weights * head_mask
+
+        attn_output = torch.matmul(attn_weights, v)
+
+        outputs = (attn_output,)
+        if output_attentions:
+            outputs += (attn_weights,)
+        return outputs
+
+    def forward(
+        self,
+        hidden_states,
+        layer_past=None,
+        attention_mask=None,
+        head_mask=None,
+        use_cache=False,
+        output_attentions=False,
+    ):
+        query = self.q_proj(hidden_states)
+
+        if layer_past is not None:
+            past = layer_past[0]
+            key_value_hidden_states = torch.cat([past, hidden_states], dim=1)
+            past_length = past.size()[1]
+        else:
+            key_value_hidden_states = hidden_states
+            past_length = 0
+
+        key = self.k_proj(key_value_hidden_states)
+        value = self.v_proj(key_value_hidden_states)
+
+        # compute block length and windows
+        bs, seq_len = hidden_states.shape[:2]
+        full_seq_length = seq_len + past_length
+        block_length = self.window_size
+        while full_seq_length % block_length != 0:
+            block_length -= 1
+        windows = full_seq_length // block_length
+
+        # create buckets
+        if layer_past is not None:
+            # we just need 1 window with block_length 1 when caching is enabled
+            query = self.create_buckets(query, 1, 1)
+        else:
+            query = self.create_buckets(query, windows, block_length)
+
+        key = self.create_buckets(key, windows, block_length)
+        value = self.create_buckets(value, windows, block_length)
+
+        key = self.look_around(key, block_length, self.window_size)
+        value = self.look_around(value, block_length, self.window_size)
+
+        # select key/value vectors only for the last window
+        if layer_past is not None:
+            key = key[:, -1:, ...]
+            value = value[:, -1:, ...]
+
+        query = self.split_heads(query)
+        key = self.split_heads(key, k=True)
+        value = self.split_heads(value)
+
+        mask = self.create_attention_mask(bs, full_seq_length, windows, block_length, attention_mask)
+        if layer_past is not None:
+            mask = mask[:, -1:, :, -1:, :]  # only take the mask for the last window
+        mask = mask.to(hidden_states.device)
```

Comment:
I think the mask should have the correct device when creating it in the function